### ISSUE TITLE
Dependency predicates

### ISSUE THREAD

**Issue by essobedo** [quarkusio/quarkus#27087]
### Description

In Quarkus, we have the notion of conditional dependency allowing to activate or not an extension if one or more extensions are present which is good enough for many different use cases but not for more complex uses cases.
Indeed, in Camel Quarkus, we would like to be able to enable the JMX extension only in JVM mode and the Debug extension only in dev mode which is not supported by the conditional dependency feature.

### Implementation ideas

The idea is to propose a new feature called "Dependency predicates" allowing to enable an extension based on criteria like the existence of an artifact but also potentially the mode used and/or the value/existence of a property. We could then support use cases like "enable this extension only in dev or test mode if and only if the extension X is present and the property `B` has been set to `foo`".

The configuration could look like this:

```
<plugin>
   <groupId>io.quarkus</groupId>
   <artifactId>quarkus-extension-maven-plugin</artifactId>
   <version>${quarkus.version}</version>
   <executions>
     <execution>
       <phase>process-resources</phase>
         <goals>
           <goal>extension-descriptor</goal> <2>
         </goals>
         <configuration>
           <dependencyPredicate> <3>
              <predicate>org.acme:quarkus-extension-c,!dev</predicate> <4>
            </dependencyPredicate>
         </configuration>
      </execution>
   </executions>
</plugin>
```

<1> runtime Quarkus extension artifact ID, in our example `quarkus-extension-b`;
<2> the goal that generates the extension descriptor which every Quarkus runtime extension project should be configured with;
<3> configuration of the predicates which will have to be satisfied for this extension to be included into a Quarkus application expressed as a list of artifacts that must be present among the application dependencies;
<4> the extension is activated only in `dev` mode if `quarkus-extension-c` is a runtime dependency.

The supported types of predicate are the following:

* `test`, `dev`, `development`, `jvm` and `native` as expected modes
* `property-name?` as test on the presence of the property `property-name`
* `property-name=expected-property-value` as test on the value of the `property-name` that is expected to be `expected-property-value`
* `groupId:artifactId[:<classifier>:<extension>]` as test on the presence of an artifact

In case a predicate is prefixed by the character `!`, the negation of the predicate is used.

**NB:** We cannot simply improve the conditional dependency feature because the conditions are combined with a logical `AND` while in this proposal the predicate entries are combined with a logical `OR` and the predicates inside a predicate entry are combined using a logical `AND` allowing to support any combinations like for example A OR (B AND C).

**Comment by essobedo** [comment]
A PR proposal https://github.com/quarkusio/quarkus/pull/27089

**Comment by essobedo** [comment]
@aloubyansky Hi, WDYT of this feature request and corresponding PR? Thx in advance 🙏

**Comment by aloubyansky** [comment]
@essobedo we have to be careful here. AFAIU, you want to enhance user experience at the end by enabling features (actually dependencies) depending on modes and environment. The issue is that it could quickly get overwhelmingly complicated for users and even us, as developers, to analyze how the final dependency tree got resolved. In addition, it will further deviate from the underlying build system (Maven/Gradle) resolver and the dependency analyzing tools built on top of that.

Could Maven profiles be used to cover some of the use-cases you mentioned? Users could also explicitly configure dependency scopes for extensions in the POM. This is something people are already familiar with and non-Quarkus tools can handle.

Do you have a use-case in mind that can't be covered using the already available config options? I guess we'll be comparing what is more complicated: using the currently available config options or adding more magic to the Quarkus dependency resolver.

Thanks.

**Comment by essobedo** [comment]
@aloubyansky thx for your quick answer.

I understand your point of view regarding the fact that it deviates from the underlying build system but as it is already more or less the case with the introduction of conditional dependency (which is an awesome idea BTW) I thought that it could still be interesting to go a little bit further to avoid being limited to what Gradle/Maven propose out of the box in term of dependency resolution.

The use case is the following: 

In Camel Quarkus, we would like to have the extension camel-debug enabled when launching Camel Quarkus in dev mode but camel-debug depends on camel-management (JMX) and either `camel-quarkus-xml-jaxb` or `camel-quarkus-xml-io-dsl`. 

The idea was to have somehow Camel Quarkus that have 3 conditional dependencies: 
* `camel-management` if JVM mode is enabled and `org.apache.camel.jmx.disabled` has been set to true
* `camel-debug` if dev mode is enabled
* `camel-quarkus-xml-io-dsl` if dev mode is enabled and `camel-quarkus-xml-jaxb` is absent

Related ticket https://github.com/apache/camel-quarkus/issues/3876

Of course, this could be managed using maven profiles but it has to be added by the end user explicitly/manually while we wanted to propose an automatic solution to improve the UX in a Framework like Camel Quarkus.

Do you see a better way to do that? Or for you the manual way is the only acceptable way?

**Comment by aloubyansky** [comment]
I think it'd be reasonable to let users decide whether they want to enable `camel-debug` in their projects by letting them add it as a `provided` dependency. That way it will be enabled in dev and test modes.
Then you could consider using `BuildStep.onlyIf` to execute only the relevant steps and skip the rest depending on a condition (launch mode, property, presence of a capability or dependency). Could that work?

**Comment by essobedo** [comment]
Regarding the conditional dependencies # 1 and # 2, I see more or less how it could be done using `BuildStep.onlyIf` but the main problem is the last one (`camel-quarkus-xml-io-dsl` should only be added if `camel-quarkus-xml-jaxb` is absent), in that case I don't see how it could be managed with `BuildStep.onlyIf`, or maybe I miss something?

**Comment by aloubyansky** [comment]
No, that can't be done from build steps. Unfortunately, there is no convenient way of doing that. Even your proposal doesn't help here. What you could explore is a possibility of extracting the code that you want to be executed by default (from the 'camel-quarkus-xml-io-dsl`) to some artifact and make `camel-debug` and `camel-quarkus-xml-io-dsl` depend on that. However, in `camel-debug-deployment` you could do some classpath analysis and load and execute that default code only if necessary.

**Comment by essobedo** [comment]
Why do you believe that my proposal doesn't help here? With my proposal, the condition would be something like `!camel-quarkus-xml-jaxb,dev` on `camel-quarkus-xml-io-dsl`

**Comment by aloubyansky** [comment]
Activating dependencies based on absence of other dependencies will lead to a confusion in some cases. Because when we activate a certain dependency based on absence of another dependency the dependency we activated may bring in some dependencies transitively that could satisfy conditions for other dependencies that would make the condition that enabled the first dependency we enabled invalid.
Another case could be when multiple conditional dependencies where found whose conditions based on absence of some dependencies could be satisfied but enabling one of those would invalidate the conditions for others. In this case, how would we choose which one to activate? And make it predictable for the users and us as developers.
These kind of issues aren't present in the impl we have today.
